### DISCUSSION TITLE
SEB requires re-login of students

### DISCUSSION THREAD

**Discussion by NoaVind** [SafeExamBrowser/seb-win-refactoring#435] [Q&A]
Hello everybody,
I am a moodle developer trying to integrate SEB into our system. As for now, when our students click "Launch Safe Exam Browser" the SEB window appears alright, but requires them to log in again into their moodle account... is there a way to pass the login cookie to SEB, allowing them to proceed without re-loggin? I chose "Configure SEB from template" when setting up the quiz, if that helps.
Thanks very much in advance,
Aiden

**Comment by danschlet** [comment]
Currently a second login in SEB is always necessary, unless you would start SEB with a generic configuration directing students to the Moodle start/home page. This configuration would need to allow reconfiguring and persist session cookies "Clear browser session when ending an exam or terminating SEB" = false (see settings in [Exam pane](https://safeexambrowser.org/windows/win_usermanual_en.html#ExamPane)). When students navigate to and start a quiz with "Launch Safe Exam Browser", then the quiz configuration is loaded and SEB reconfigured accordingly. In this use case each quiz configuration needs to set the setting "Clear browser session when starting an exam or starting SEB" = false. Then the Moodle user login is persisted from the initial SEB start with the generic config to the reconfiguring for the quiz config.

For transferring the Moodle session cookie from a third party browser to SEB as in your use case, the Moodle-SEB integration would need to be improved. For example a modified version of the Moodle auto login feature could be used (which generates a one-time auto login URL for a specific user, afaik used in the Moodle app). For that you would need to place a feature request on moodle.org, as we are not developing the SEB integration in Moodle ourselves.

**Reply by NoaVind** [reply]
okay, I understand now. 
Thank you very much, you really helped me

**Reply by AntonT76** [reply]
Dear @danschlet

we even need a third login in SEB - how can we reduce it to 2-time-login?
We also uses a template and generate the template directly in the SEB configuration tool and upload the seb file as template. 
1. Thirst we have to login to the examination Moodle
2. Press button "launch Safe Exam Browser" - SEB will be loaded and I will be forwarded to the login page of the examination Moodle
3. type in login credentials - I will be forwardet again to the login page
4. type in login credentials again
5. start quiz

Any ideas whats wrong? It just happens if I upload seb-file which will be created by the SEB configuration tool. If I use manual configuration or upload a seb-file which will be created directly in Moodle, we just need 2times login.

br, Anton

**Reply by danschlet** [reply]
Are you using the current SEB versions? This should not happen and sounds like a bug. Can you provide a testing login where we can reproduce that?

**Reply by AntonT76** [reply]
we are using current SEB version. I can provide you a testing login on our Moodle testing system (you will get 1 test course and a specific local test user). If it's ok for you I can send you this information to your ethz-address in the next days.
br Anton

**Reply by danschlet** [reply]
Yes, please send us that information, either if you have the ethz address or info at safeexambrowser.org.

**Reply by 1JackBlack1** [reply]
This is due to how Moodle launches SEB when you click the "Launch Safe Exam Browser" button.
And to me this happens regardless of if you upload a config or use one Moodle has made.
The more detailed process is:
1 - Student logs into Moodle and navigates to the quiz page.
2 - Student presses button to launch SEB. This is a sebs:// type url which tells SEB to launch and go to a particular link.
3 - SEB loads (using the client config) and navigates to that page.
4 - Moodle sees them as a guest and requires them to log in.
5 - After they login to Moodle, the SEB config file loads and is applied.
6 - SEB applies the config file which wipes the session and goes to the specified start URL.
7 - With their new session, Moodle sees that they are not authenticated and requires them to log in.

If instead you download the config and launch it, SEB will directly launch with that config file.

The issue is in the function get_seb_config_content, in /mod/quiz/accessrule/seb/classes/helper.php on line ~121.
This calls require_login, which means the user needs to have logged into Moodle to access the page. Otherwise, anyone could hypothetically get the config file without needing to authenticate.

**Reply by danschlet** [reply]
Step 7 should not happen (and doesn't in our testings), because SEB doesn't wipe the session after applying the config file in step 6. Does this still happen for you? With current SEB versions for Windows? With SEB for macOS or iOS?
